### PR TITLE
deploy Calico API Server with soft zone based anti-affinity by default

### PR DESCRIPTION
**Pull request by calvinrzachman**
This PR deploys the Calico API Server component with zone based pod anti-affinity by default for deployments exceeding a single replica. This is "preferred" during scheduling so will NO-OP for single zone clusters (ie: those with a single `topology.kubernetes.io/zone: X` label).

I have built a development version of the image using the `v1.27` release branch + this commit and deployed it to a RedHat OpenShift 4.10 cluster running on IKS using Calico v3.23.1 (OS). I observe that the `calico-apiserver` deployment yaml is structured as expected with both zone and host pod anti-affinity rules.

I have never used Ginkgo before, but was able to modify the tests to cover the newly added pod affinity and get all tests passing.`make test` was presenting some problems, which I have yet to be able to debug. However the same error also presents when run against `master` so it seems unlikely to do with this change. Let me know what more testing is needed if any.

Thanks!

Addresses: IBM request for increased HA of components deployed via Tigera Operator, via sensible K8s deployment API defaults.

Used https://github.com/tigera/operator/pull/1937 as a guide.

Affects: Calico API Server (main concern), Tigera Operator |  dex, logStorage, manager, esGateway (NOTE: These are probably Enterprise components. I have never used these so get the a-okay to add anti-affinity for these.)

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Comment by tmjd** [comment]
Thank you for submitting this. I think an improvement on this would be to update the function `NewPodAntiAffinity` in  `pkg/render/common/podaffinity/pod_anti_affinity.go` and add the zone anti-affinity that way, instead of customizing the anti-affinity for apiserver.
WDYT?

If we change that function, then all the other places we use that function would pick up that feature/change also. I did a grep to see the other places that use that function and I don't think there is any reason it would be a bad idea.
I think most of the changes you needed to make to the test code would not be needed either.

**Comment by tmjd** [comment]
BTW I triggered another run of CI because I'm not sure why there was a failure, I think it was probably a flake. I don't think your changes would have had any impact on the test, so lets see.

**Comment by tmjd** [comment]
The most recent run failed due to an unrelated update that was needed. I've put up #2053 with the needed update, after it is merged could you rebase your changes on the master branch, then force push your changes, and that should trigger CI to run again. If you need help with rebasing let me know and I can give you the commands.

**Comment by calvinrzachman** [comment]
Hi Erik, thanks for the review. That is a good idea. I must have convinced myself against that before - maybe I just thought I'd limit my change to impact just the `apiserver` lest I run into problems configuring this for other components. I'll take another look at updating `pod_anti_affinity.go` and see if I can get things working that way.

**Comment by calvinrzachman** [comment]
Updated PR to configure zone based pod anti-affinity via `pkg/render/common/podaffinity/pod_anti_affinity.go` instead of customizing the anti-affinity for `apiserver` component only.

Previous method of changing API server only can be found here: https://github.com/calvinrzachman/operator/commit/9c87df13a3893697cb54a776f7c3400a67a8ca3f

File: pkg/render/common/podaffinity/pod_anti_affinity.go
Comment (calvinrzachman) [reply pkg/render/common/podaffinity/pod_anti_affinity.go:49]: Based on the [K8s docs](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) I don't think it's required for the case where instances of a pod are taking other instances of the *same* pod into account during scheduling. It sounds strictly necessary when you seek to schedule relative to pods in other namespaces. 

> You express these rules (Y) as [label selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors) with an optional associated list of namespaces. Pods are namespaced objects in Kubernetes, so Pod labels also implicitly have namespaces. Any label selectors for Pod labels should specify the namespaces in which Kubernetes should look for those labels.

> In addition to labelSelector and topologyKey, you can optionally specify a list of namespaces which the labelSelector should match against using the namespaces field at the same level as labelSelector and topologyKey. If omitted or empty, namespaces defaults to the namespace of the Pod where the affinity/anti-affinity definition appears.

They say it's optional, say you "should" use it, then say that it defaults to the namespace the definition appears in. I don't see a problem in adding it though.

**Comment by lmm** [comment]
In case you haven't seen it, we're working on some operator enhancements to improve the component resource config options. Relevant to this PR is that it will allow users to specify Calico API server's affinity (and more). See https://github.com/tigera/operator/issues/1990 for details.

In a previous version of this PR there was a TODO that we should allow user configuration so I thought I'd mention it :)

**Comment by calvinrzachman** [comment]
@lmm Thanks for the heads up there Laurence! Do you think it's okay to merge this PR once ready and let that work add the user defined configuration to build on top of this? This PR is just in the interest of a quick and sensible default which will help improve HA in multi-zone environments, understanding that modifying the CRDs to enable user configuration is a bit more work 😄

**Comment by lmm** [comment]
@calvinrzachman Yes absolutely, I think you're right that this PR should go in first and the user-provided config changes will come after 👍

**Review by tmjd — APPROVED** [review]
LGTM

**Comment by tmjd** [comment]
I'm going to close then re-open this to trigger CI

**Comment by tmjd** [comment]
The failure in CI was not related to the changes here, it was in an error because of an updated that is needed in master. The FV and UTs passed so I'm going to merge this.
@calvinrzachman thank you for the contribution.

**Pull request merged**
